src/navmesh.js

```diff
@@ -373,98 +373,113 @@ export class NavMesh {
+                newPoint = portal.p1;
+            } else {
+                throw new Error("Invalid portal: Funnel cannot be extended on both sides at once.");
             }
 
-            edge1 = newEdge1;
-            edge2 = newEdge2;
-
-            // If we are at the end of the path,
-            // just add the last edge, which is from -> to
-            if (i === path.length - 2) {
-                const toEdge = new Edge(from, to);
-                let newToEdge = this._funnelEdge(edge1, edge2, toEdge);
-                if (newToEdge === edge1) {
-                    points.push(edge1.p2);
-                } else if (newToEdge === null) {
-                    points.push(edge2.p2);
-                }
+            // Extend funnel
+            this._extendFunnel(tail, left, right, extendLeft, newPoint);
+        }
+
+        // Close funnel to endpoint
+        this._extendFunnel(tail, left, right, true, to);
+        this._extendFunnel(tail, left, right, false, to);
+
+        return tail;
+    }
 
-                points.push(to);
+    _extendFunnel(tail, left, right, extendLeft, newPoint){
+        const apex = tail[tail.length - 1];
+        // We pretend to be in the `expandLeft` case here. Otherwise flip.
+        if (!extendLeft) {
+            [left, right] = [right, left];
+        }
+        // Determine angle of `apex`-`newPoint` relative to `apex`-`left[j]`
+        let j = 0;
+        while (j < left.length
+                && this._isInLeftRightOrder(apex, newPoint, left[j], !extendLeft)) {
+            j++;
+        }
+        // All points in `left` with index `< j` are right of `newPoint` and
+        // all points in `left` with index `>= j` are left of or at the same angle as `newPoint`. 
+        left.length = j; // Shrink funnel if `j < left.length`
+        left.push(newPoint);
+        if (j === 0) {
+            // If the funnel shrunk all the way on the left, it might collapse to the right.
+            // Determine how far it needs to collapse
+            let k = 0;
+            while (k < right.length
+                    && !this._isInLeftRightOrder(apex, newPoint, right[k], !extendLeft)) {
+                k++;
             }
+            // All points in `right` with index `< k` are left of or at the same angle as `newPoint` and
+            // all points in `right` with index `>= k` are right of `newPoint`. 
+            tail.push(...right.splice(0, k)); // Collapse funnel if `k > 0`
         }
+    }
 
-        return points;
+    /** 
+     * Are the points `p1` and `p2` in left-to-right order, viewed from `origin`?
+     * Checks for right-to-left order instead if `flip = true`.
+     * Returns `false` if the angles are equal.
+     */
+    _isInLeftRightOrder(origin, p1, p2, flip = false) {
+        if (flip) {
+            [p1, p2] = [p2, p1];
+        }
+        const vec1 = p1.sub(origin);
+        const vec2 = p2.sub(origin);
+        return cross(vec1, vec2) > 0;
```

Comment:
I also tested this and it seems that if you change the sign, all test pass anyways (the algorithm is invariant if you do it clockwise or counter-clockwise). I think we need to swap the sign here, because we have a function that is called '_isInClockwiseOrder' which actually checks if it is in counter-clockwise order :-)

Reply:
I think the right-hand rule is for mathematical coordinate systems, which have the y-axis pointing up. But we have the y-axis pointing down. That's why everything is inverted. Just calculate a simple example like (0,1) and (1,0) and you'll see that the current implementation is correct.

Reply:
We're talking about a mathematical coordinate system. Nowhere in the library is there the assumption that the Y axis is pointing down. 

The fact that you then use it in a game with Y pointing down has the only effect of flipping the entire navmesh, so that the logic is still correct (as if you saw the thing on a mirror).

I would prefer to keep everything with the same standard math definition so that it is easier to reason about.

Reply:
Ok, sure.